src/drivers/magnetometer/memsic/mmc5983ma/mmc5983ma.cpp

```diff
@@ -0,0 +1,230 @@
+			write_register(MMC5983MA_ADDR_CTRL_REG0, MMC5983MA_CTRL_REG0_TM_M | set_reset_flag);
+
+			_collect_retries = 0;
+			_state = State::Collect;
+
+			auto elapsed = hrt_elapsed_time(&start_time);
+
+			if (elapsed > 5_ms) {
+				ScheduleNow();
+
+			} else {
+				ScheduleDelayed(5_ms - elapsed);
+			}
+
+			return;
+		}
+
+	case State::Collect: {
+
+			auto start_time = hrt_absolute_time();
+
+			uint8_t status = read_register(MMC5983MA_ADDR_STATUS_REG);
+
+			if (status & MMC5983MA_STATUS_REG_MEAS_M_DONE) {
+				SensorData data = {};
+
+				if (read_register_block(&data) != PX4_OK) {
+					PX4_DEBUG("read failed");
+					perf_count(_comms_errors);
+					_state = State::Measure;
+					_sample_index = 0;
+					ScheduleDelayed(100_ms);
+					return;
+				}
+
+				// Measurement available
+				_measurements[_sample_index] = data;
+				_sample_index++;
+
+				if (_sample_index > 1) {
+					publish_data();
+					_sample_index = 0;
+					perf_count(_sample_perf);
+				}
+
+				_state = State::Measure;
+
+				auto elapsed = hrt_elapsed_time(&start_time);
+
+				if (elapsed > 5_ms) {
+					ScheduleNow();
+
+				} else {
+					ScheduleDelayed(5_ms - elapsed);
+				}
+
+				return;
+
+			} else {
+				PX4_DEBUG("not ready");
+				perf_count(_comms_errors);
+				_collect_retries++;
+				_state = _collect_retries > 3 ? State::Measure : State::Collect;
+				ScheduleDelayed(5_ms);
+				return;
+			}
+		}
+	} // end switch/case
+}
+
+void MMC5983MA::publish_data()
+{
+	uint32_t xraw_1 = (_measurements[0].xout0 << 10) | (_measurements[0].xout1 << 2) | ((
+				  _measurements[0].xyzout2 & 0b11000000) >> 6);
+	uint32_t yraw_1 = (_measurements[0].yout0 << 10) | (_measurements[0].yout1 << 2) | ((
+				  _measurements[0].xyzout2 & 0b00110000) >> 4);
+	uint32_t zraw_1 = (_measurements[0].zout0 << 10) | (_measurements[0].zout1 << 2) | ((
+				  _measurements[0].xyzout2 & 0b00001100) >> 2);
+
+	uint32_t xraw_2 = (_measurements[1].xout0 << 10) | (_measurements[1].xout1 << 2) | ((
+				  _measurements[1].xyzout2 & 0b11000000) >> 6);
+	uint32_t yraw_2 = (_measurements[1].yout0 << 10) | (_measurements[1].yout1 << 2) | ((
+				  _measurements[1].xyzout2 & 0b00110000) >> 4);
+	uint32_t zraw_2 = (_measurements[1].zout0 << 10) | (_measurements[1].zout1 << 2) | ((
+				  _measurements[1].xyzout2 & 0b00001100) >> 2);
+
+	// NOTE: Temperature conversions did not work
+	// float trawf = float(_measurements[0].tout + _measurements[1].tout) / 2.f;
+	// float temp_c = trawf * 0.8f - 75.f;
+	// _px4_mag.set_temperature(temp_c);
+
+	// +/- 8 Gauss full scale range
+	// 18-bit mode scaling factor: 0.0625 mG/LSB
+	float x1 = 8.f - (float(xraw_1) * 0.0625f) / 1e3f;
+	float x2 = 8.f - (float(xraw_2) * 0.0625f) / 1e3f;
+	float y1 = 8.f - (float(yraw_1) * 0.0625f) / 1e3f;
+	float y2 = 8.f - (float(yraw_2) * 0.0625f) / 1e3f;
+	float z1 = 8.f - (float(zraw_1) * 0.0625f) / 1e3f;
+	float z2 = 8.f - (float(zraw_2) * 0.0625f) / 1e3f;
+	float x = (x1 - x2) / 2.f;
+	float y = (y1 - y2) / 2.f;
```

Comment:
Y needs to be inverted.

![image](https://github.com/PX4/PX4-Autopilot/assets/2019539/88df1676-90bf-4853-9bc1-181afc7483f5)